File: trenscoda.rb
require 'openai'
require 'dotenv/load'
require 'json'
require 'fileutils'
require 'open3'
require 'colored'
require 'diffy'

OpenAI.configure do |config|
  config.access_token = File.read('openai_key.txt').strip
end

def send_code_to_gpt(client, script_name,language1,language2)
  file_lines = File.readlines(script_name)
  file_with_lines = file_lines.each_with_index.map { |line, idx| "#{idx + 1}: #{line}" }.join

  initial_prompt_text = File.read('prompt.txt')

  prompt = build_prompt(initial_prompt_text, file_with_lines,language1,language2)

  response = client.chat(
    parameters: { model: 'gpt-3.5-turbo', messages: [{ role: 'user', content: prompt }], temperature: 1.0 }
  )

  output =  response.dig('choices', 0, 'message', 'content').strip
  status = response.dig('object')
  [output, status]
end

def build_prompt(initial_prompt_text, file_with_lines,language1,language2)
  initial_prompt_text +
    "\n\n" \
    "Here is the code that needs to convert from "+language1.to_s+" to "+language2.to_s+":\n\n" \
    "#{file_with_lines}\n\n" \
    'Please provide your suggested changes'
end

def main
  if ARGV.length < 2
    puts 'Usage: trenscoda.rb <file_name> <language1> <language2> ...'
    puts 'Remember that file_name is written in language1'
    exit(1)
    end

    file_name = ARGV[0]
    language1 = ARGV[1]
    language2 = ARGV[2]
    client = OpenAI::Client.new

  loop do
    output, status = send_code_to_gpt(client,file_name,language1,language2)

    if status == 'chat.completion'
      puts 'Script ran successfully.'.blue
      puts "Output: #{output}"
      break
    else
      puts 'Script crashed. Trying to fix...'.blue
      puts "Output: #{output}"
    end
  end
end

main if __FILE__ == $PROGRAM_NAME

File: trenscoda.py
import openai
import sys
import os
import json
import shutil
import subprocess
from difflib import unified_diff

openai.api_key = open('openai_key.txt').read().strip()

def send_code_to_gpt(script_name,language1,language2):
    with open(script_name) as f:
        code = f.read()

    file_lines = code.split('\n')
    file_with_lines = '\n'.join([f'{i+1}: {line}' for i, line in enumerate(file_lines)])

    initial_prompt_text = open('prompt.txt').read()
    prompt = build_prompt(initial_prompt_text, file_with_lines,language1,language2)

    response = openai.ChatCompletion.create(
        model="gpt-3.5-turbo",
        messages=[{"role": "user", "content": prompt}],
        temperature=1.0,
    )

    output = response.choices[0].message.content
    status = response.object

    return output, status

def build_prompt(initial_prompt_text, file_with_lines,language1,language2):
    return initial_prompt_text + '\n\n' \
        f'Here is the code that needs to convert from {language1} to {language2}:\n\n' \
        f'{file_with_lines}\n\n' \
        'Please provide your suggested changes'

def main():
    if len(sys.argv) < 3:
        print('Usage: trenscoda.py <file_name> <language1> <language2> ...')
        print('Remember that file_name is written in language1')
        exit(1)

    file_name = sys.argv[1]
    language1 = sys.argv[2]
    language2 = sys.argv[3]

    while True:
        output, status = send_code_to_gpt(file_name,language1,language2)

        if status == 'chat.completion':
            print('Script ran successfully.')
            print(f'Output: {output}')
            break
        else:
            print('Script crashed. Trying to fix...')
            print(f'Output: {output}')

if __name__ == '__main__':
    main()
